### ISSUE TITLE
Improve support for message batch processing

### ISSUE THREAD

**Issue by rbeckman-nextgen** [nextgenhealthcare/connect#3162]
Imported Issue. Original Details:
Jira Issue Key: MIRTH-3251
Reporter: narupley
Created: 2014-04-28T09:12:54.000-0700

**Comment by rbeckman-nextgen** [comment]
If we allow messages to be processed as if it were a batch message, there is no longer guaranteed delivery. Currently we only support batch on File Receiver and TCP Receiver

If something goes wrong with the File Receiver, the file will not be moved/deleted so it will be picked up again when the channel starts. Any messages in the batch that had already processed will be processed again.

On the TCP Receiver, an ack is sent back after the first message. Once the first message in the batch is received, subsequent messages in the batch will be lost if the server goes down.

With the additions I'm making that will allow messages that are read from any connector to be batch processed, the message may be read into memory completely first before being split into a batch. Once again, other than the first message, if the server goes down none of the remaining messages will be recovered.

I don't see this as a big problem yet because its something that we weren't exactly handling to begin with. If we did want to guarantee delivery with batch, it would require persisting the original message somehow. Even then it would only work for messages that are fully cached first. Batch messages that are streamed are dependent on the connector logic on how they handle failure. For simplicity's sake, I'm simply going to ignore this problem for now.

Imported Comment. Original Details:
Author: wayneh
Created: 2014-06-16T17:40:34.000-0700

**Comment by rbeckman-nextgen** [comment]
The TCP Listener **currently** sends a response back after the first message, but the idea is to allow the response to be sent back at the very end after all messages have finished. So in general, this will/should be handled with MIRTH-3252 and the ability to have access to batch metadata (MIRTH-1800, MIRTH-2132, MIRTH-2428). Following the HL7 specifications, responses to batch messages should usually be batch messages themselves, containing ACKs for each message in the original batch. The user should have some way to build up a batched response message-by-message, and then send it off at the very end. With the aforementioned issues that could technically be done with the global channel map or something, but hopefully there's an easier solution for the user.

If a batch response is only being sent back after the _last_ message and the connection goes down in the middle, then obviously no ACK will be sent back, and the originating system will be expected to resend the message. Or in the case of the File Reader, the file should only be moved/deleted after the _last_ message finishes, so the file should be read in again. That's fine; as long as we include a sufficient amount of metadata in the source map for each batched message, users will be able to do things like detect duplicate messages and act accordingly.

For HL7 v2.x (or even for all data types in the case of a CRC), we _could_ add some sort of convenience property like "automatically detect and discard duplicate messages" using the MSH.10 or a CRC, and some sort of sliding window. That could be useful not only for batching but for processing in general (there have been several "how do I ignore duplicate HL7 messages" threads). I can already hear the cries of "but that won't work **100%** of the time!", and that's true, but let's not allow Perfect to be the enemy of Good...

Imported Comment. Original Details:
Author: narupley
Created: 2014-06-16T18:35:40.000-0700

**Comment by rbeckman-nextgen** [comment]
Batch support has been added for most data types and connectors. For most data types we've added a JavaScript batch script so users can split their own messages. For others there are some built in methods of splitting batch, such as by element for XML. In the future we may be adding more built in methods of splitting for various data types.

In addition each connector can now process batch messages. Some connectors can stream the message directly from the source, while others need to read the entire batch into memory first. In the future we may be adding streaming for more connectors/data type combinations.

Imported Comment. Original Details:
Author: wayneh
Created: 2014-08-05T20:01:09.000-0700
